### DISCUSSION TITLE
Communicating breaking API changes

### DISCUSSION THREAD

**Discussion by marcus-at-localhost** [atrocore/atrocore#822] [Ideas]
After updating from PIM 1.9.12 to 1.9.31 I experienced another breaking change in the API that a 3rd party client of ours consumes.

I'm following the changelog very closely via email alerts and rss (eg https://github.com/atrocore/atropim/releases.atom).
Occasionally there is an additional note in the changelog, and of course there are the diffs of all changes. That's good and transparent, but I'm not knee-deep in the codebase and also not full-time working with the PIM, so often it doesn't mean anything.

With my latest update the access or relationship to `/products/:id/assets` broke. So I go through the changelog notes and there is no information about potential API changes. I remembered you were talking about having a unified changelog soon, and I found it on your help page: https://help.atrocore.com/release-notes/atropim/1.9/. Ok. 

> AtroPIM 1.9.21
> From: 2023-08-01
> Refactored relationship entities

that's a hint, but in no way obvious, that this breaks API endpoints.

Here are my ideas or wishes:

- More detailed changelog on github release page, and especially with a note about breaking API changes (anytime you have to change endpoints yourself internally). Maybe knowing the reasoning behind the change would help as well. (I know documentation is a lot of additional work)
- Access to the changelog from within the Modul Update page in the PIM Admin (link or just integrated).
- Maybe there can be a community effort to document API changes. I'm happy to start a thread here where others can chime in and contribute because I'm only focus on certain aspects of the API myself. That way we (or maybe only I) don't need to open a new thread every time a breaking change occurs. 

What do you think? (But more detailed changelogs would be great)

Thanks, Marcus

PS: while I wrote this, 2 version changes happened :D

**Comment by atrolex** [comment]
Hello Marcus,

we try to ensure backward compatibility as long as possible. But here we always have a problem with optimization and maintenance of what is no longer relevant, it cost time and resources as well. Therefore, we can promise that we will monitor such changes more closely and better describe these things in the release notes, this is now the task for our QA engineers. As for the display of release notes in the manager module, this task is already in our backlog. As for the community, here we are completely open and everyone can contribute as they wish.

Please note, we are dropping support for github release notes functionality, look for the most updated versions here:
https://help.atrocore.com/release-notes/
or, alternatively (just change the name of module and version to get the info):
https://rn.atrocore.com/atrocore/1.6.md

I have also added RSS:
![image](https://github.com/atrocore/atrocore/assets/33658481/99b79e9e-6f5d-4ba7-9979-ce8889d4db7f)

**Reply by marcus-at-localhost** [reply]
@atrolex I appreciate your reply and insight and completely understand that there are reasons for you and the team, to change API endpoints. It's really just about getting a heads-up for potentially breaking changes or removal of endpoints, and maybe a hint on how to achieve a previous behavior or so. Of course, the API is complex and most of the time I don't even notice changes because I work only with a fraction of what the API has to offer. Well, we will see. I'm positive it's going to be fine with you providing a more detailed changelog and with me having a more robust system that detects breaking changes more easily (like using a Data Transfer Object).
Thank you so much.
